main/src/main/scala/sbt/internal/server/BuildServerReporter.scala

```diff
@@ -188,6 +204,10 @@ final class BuildServerReporterImpl(
   }
 }
 
+object BuildServerReporterImpl {
+  private val hasAnyProblems = mutable.Set[Path]()
```

Comment:
Is this thread-safe, i.e. is it guaranteed that this global mutable data structure will be accessed only from 1 thread ever?

Reply:
Thanks for this question @asflierl, I was wondering about this too but I forgot to ask about it. I think there can be only one compilation at time for the build target, but multiple build targets can be compiled in parallel.  
What if one file belongs to 2 build targets (it's very common to have multiple build targets for multiple scala versions) and there is problem in 2.13, but no problem in 2.12. If we're using just paths then we will end up with race conditions which build target will clear. 

I'll need to think of these concerns.

Reply:
Yes indeed, this is a important concern but it seems we don't have a problem with the current implementation.

There is one instance of `BuildServerReporterImpl` for each build target and, as you said @kpodsiad, there can be only one compilation at time for the build target. So it seems there cannot be any race condition.

Reply:
> There is one instance of `BuildServerReporterImpl` for each build target and, as you said @kpodsiad, there can be only one compilation at time for the build target. So it seems there cannot be any race condition.

@adpi2 
Are you sure on part `There is one instance of BuildServerReporterImpl for each build target`? Because when I was playing with it it seemed like each compilation gets fresh `BuildServerReporterImpl` instance. I've added `hasAnyProblems` to the class and I was printing all problems in `reportSucess` and `reportFailure`. And even though problems were set in `reportFailure`, `reportSucess` was always printing an empty set.   
My workflow was following:
- introduce error in file (problems were printed)
- fix error (I tried to print problems before clearing them but there weren't any)

That's why I added `hasAnyProblems` to the companion object and in this case we might have race conditions because `hasAnyProblems` isn't scoped to one build target.

Reply:
Additionally, even if there was only one instance of the reporter accessing the global object (and therefore no race conditions), the issue of visibility remains: if the reporter is executed on a thread pool, it is not guaranteed that the same thread is used for every call to the reporter, so there is no clear "happensBefore" relation established between such calls, which might corrupt the internals of the mutable hash set.

Reply:
You don't even need to add a key, you can use an anonymous `Def.setting` instead, like that:

```scala
private def bspCompileState: Initialize[BspCompileState] = Def.setting {
  new BspCompileState()
}
```
where `BspCompileState` is a class that contains the mutable set of a single build target.

Then in the `bspReporter` task:

```scala
new BuildServerReporter(..., bspCompileState.value, ...)
```

`bspCompileState` is a setting so it will be the same instance on each run and it will be automatically scoped to a build target (one instance for each build target).

I also suggest to move the mutable `compiledTargetsAtLeastOnce` into the `BspCompileState` instance, as a boolean.